sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -221,6 +221,46 @@ object DataSourceAnalysis extends Rule[LogicalPlan] with CastSupport {
   }
 }
 
+/**
+ * Add a repartition by dynamic partition columns before insert Datasource table.
+ *
+ * Note that, this rule must be run after `DataSourceAnalysis`.
```

Comment:
Can you describe why it should run after `DataSourceAnalysis`?

Reply:
To use `CreateDataSourceTableAsSelectCommand`, `InsertIntoDataSourceDirCommand` and `InsertIntoHadoopFsRelationCommand`.
https://github.com/apache/spark/blob/95639a763cd4f73d295c11f21701321905399787/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala#L139-L156

Reply:
That's fine but it would be great to explain here in a comment.